Make it possible to search for the source code as if the patch was applied

boiko opened the pull request:
This is useful when verifying if a new release of the software has the fix applied upstream.

boiko committed:
Start implementing a applied patch search

boiko committed:
Fix case when mode is not set

boiko committed:
Fix building

boiko committed:
Show potentially fixed code in the web UI as well

moio commented on src/main/java/com/suse/pase/query/QueryFactory.java:
```diff
@@ -40,6 +38,23 @@
+                            .filter(d -> of(INSERT, EQUAL, CHANGE).contains(d.getType()))
+                            .map(d -> d.getTarget().getLines().stream().filter(l -> !l.isEmpty()).collect(toList()))
+                            .collect(toList());
+                    return new PatchTargetQuery(path, chunks);
```
```suggestion
                    return new PatchQuery(path, chunks);
```

Alternate class name suggestion, to make clear there are two possible cases for this query (patch target or applied patch).

PS. I recommend to use an IDE to do this kind of refactoring (renaming) in a safe way.

I highly recomment IntelliJ IDEA if you are not using it already.

moio replied:
Once the class is renamed to `PatchQuery`, I'd DRY up `buildAppliedPatchTargetQuery` and `buildPatchTargetQuery` by creating a generic `buildPatchQuery` which takes a `List<DeltaType>` as parameter.

Then I can imagine implementing `buildAppliedPatchTargetQuery` by calling `buildPatchQuery` with `of(DELETE, EQUAL, CHANGE)` and `buildAppliedPatchTargetQuery` with `of(INSERT, EQUAL, CHANGE)`.

Finally I would recommend renaming `buildAppliedPatchTargetQuery` to `buildAppliedPatchQuery` as "target" does not make too much (English) sense at that point.

moio commented on src/test/java/com/suse/pase/MainTest.java:
Ideally we should add tests for the new query types - better but not necessarily in this PR, just eventually.

boiko replied:
I will try to write some in a separate pull request, let's see how that goes.

moio reviewed (commented):
It looks pretty good already! I only have a couple of suggestions which are mostly cosmetic/stylish in their nature.

Then this is ready to merge (as soon as you tested that it works! :wink:)

Good job!

boiko committed:
Rename PatchTargetQuery to PatchQuery for clarity

boiko committed:
Rewrote some methods to reduce code duplication

The patch queries were added to a generic buildPatchQuery() function
that takes a boolean controlling whether it is source or target

moio reviewed (approved):
LGTM. If even more types are ever to be added, one option is to use an enumeration instead of passing even more booleans around. But that's style nitpicking by and large!
